cpp/mrc/tests/modules/test_mirror_tap_util.cpp

```diff
@@ -0,0 +1,201 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2021-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "test_modules.hpp"
+
+#include "mrc/core/executor.hpp"
+#include "mrc/engine/pipeline/ipipeline.hpp"
+#include "mrc/modules/mirror_tap/mirror_tap_util.hpp"
+#include "mrc/modules/module_registry.hpp"
+#include "mrc/options/options.hpp"
+#include "mrc/segment/builder.hpp"
+
+#include <gtest/gtest-message.h>
+#include <gtest/gtest-test-part.h>
+
+#include <utility>
+#include <vector>
+
+using namespace mrc;
+
+TEST_F(TestMirrorTapUtil, SinglePipelineTapAndBufferTest)
+{
+    using namespace modules;
+    const std::string test_name{"SinglePipelineTapAndBufferTest"};
+
+    // Create external captures for packet counts.
+    unsigned int packet_count{10000};
+    unsigned int packets_main{0};
+    unsigned int packets_mirrored{0};
+
+    auto config = nlohmann::json();
+
+    auto init_wrapper_main = [&packets_main, packet_count, test_name](segment::Builder& builder) {
+        auto source = builder.make_source<std::string>(test_name + "_main_source",
+                                                       [packet_count](rxcpp::subscriber<std::string>& sub) {
+                                                           if (sub.is_subscribed())
+                                                           {
+                                                               for (unsigned int i = 0; i < packet_count; i++)
+                                                               {
+                                                                   sub.on_next(std::to_string(packet_count));
+                                                               }
+                                                           }
+
+                                                           sub.on_completed();
+                                                       });
+
+        auto sink = builder.make_sink<std::string>(test_name + "_main_sink", [&packets_main](std::string input) {
+            packets_main++;
+        });
+
+        // Untapped edge that will be broken and tapped by the mirror tap.
```

Comment:
I'll make this comment here, but it applies broadly.

The comment suggests that you will take an edge and perform some surgery to insert your tap. This is a big PR, so I haven't really gotten down into the details yet, so the first question is, are you performing an edge break?

If so, can you explain the rational for why you choose this method over the prologue/epilogue taps that are currently in place?

For a given edge, you could either epilogue tap on the edge which would provide you access to the outbound data object before any channel effects (if transport is over a channel and not direct). Alternatively, you could prologue tap a sink and get the data after any channel policies were enforced (e.g. RecentChannel which might drop data).

I could have missed where you are using the epilogue/prologue taps as not all the file changes get populated via the web ui, but a quick search does not show any calls to `add_prologue_tap` or `add_epilogue_tap`.

Reply:
We can discuss, part of the approach is to improve easy of use/quality of life for end users, so that their initializer code doesn't actually need to be modified and this type of workflow can be supported:

```python
    tapped_init_wrapper_main = mirror_tap.tap(init_wrapper_main,
                                              test_name + "_main_source",
                                              test_name + "_main_sink")

    tapped_init_wrapper_mirrored = mirror_tap.stream_to(init_wrapper_mirrored,
                                                        test_name + "_mirror_sink")

    pipe.make_segment("segment_main", [], [mirror_tap.get_egress_tap_name()], tapped_init_wrapper_main)
    pipe.make_segment("segment_mirror", [mirror_tap.get_ingress_tap_name()], [], tapped_init_wrapper_mirrored)```